Enhance WooCommerce Admin Performance by Splitting Scripts

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to https://github.com/woocommerce/woocommerce/issues/54992#issuecomment-2677531830

This PR refactors the WooCommerce Admin client-side architecture to improve performance and maintainability by splitting the scripts into two separate entry points:

1. WC App - For WooCommerce Admin pages
2. WC Embed - For embedded WooCommerce components in classic admin pages

Additionally, the PR includes:

- Use TypeScript to refactor the split client-side code
- Add more unit tests for related changes
- Extraction of shared logic into separate utility files to avoid duplication and loading unnecessary code
- Fix admin-layout slot fills component types as errors raised when using the components in the header.

The primary motivation for this change is to enhance performance for classic WooCommerce admin pages (orders, products, etc.). Currently, all WooCommerce admin scripts are loaded on every admin page, but embedded pages only require certain components. By splitting the scripts into separate entry points, we can significantly reduce the JavaScript payload for classic admin pages, resulting in faster page loads.

```
Before:
└── client/admin/
    └── client/
        ├── index.js                 # Single entry point for all admin pages
        ├── header/
        │   └── index.js             # Header components for all contexts
        ├── layout/
        │   └── index.js             # Layout components for all contexts
        └── stylesheets/
            └── _index.scss          # All styles bundled together

After:
└── client/admin/
    ├── webpack.config.js            # Updated with multiple entry points
    └── client/
        ├── index.tsx                # Entry point for WC Admin pages
        ├── embed.tsx                # New entry point for embedded pages
        ├── shared.tsx               # Shared utilities between both entry points
        ├── header/
        │   ├── index.tsx            # Header for WC Admin pages
        │   ├── embed.tsx            # Lightweight header for embedded pages
        │   └── shared.tsx           # Shared header components
        ├── layout/
        │   ├── index.tsx            # Layout for WC Admin pages
        │   ├── embed.tsx            # Lightweight layout for embedded pages
        │   └── shared.tsx           # Shared layout components
        └── stylesheets/
            ├── _index.scss          # Styles for WC Admin pages
            └── _embed.scss          # Minimal styles for embedded pages
```

#### Script size

Comparison of Script Sizes and Resources Loaded:

| Page   | Before: Entry Script Size | Before: Resources Loaded | After: Entry Script Size | After: Resources Loaded |
|--------|---------------------------|--------------------------|--------------------------|-------------------------|
| Home   | 211 kB                    | 8.7 MB                   | 74.1 kB                  | 8.6 MB                  |
| Order  | 211 kB                    | 9.1 MB                   | 135 kB                   | 8.6 MB                  |

Before Screenshots:
- Home: ![Screenshot 2025-02-25 at 16 36 21](https://github.com/user-attachments/assets/10df7e54-6560-4244-adb9-a27afbee216b)
- Order: ![Screenshot 2025-02-25 at 16 37 01](https://github.com/user-attachments/assets/10070738-b449-4b46-a530-9864b4a32027)

After Screenshots:
- Home: <img width="1147" alt="Screenshot 2025-02-25 at 17 17 20" src="https://github.com/user-attachments/assets/5d5a6a70-d9ed-4143-a7a8-81f674eb7f5f" />
- Order: <img width="1147" alt="Screenshot 2025-02-25 at 17 17 44" src="https://github.com/user-attachments/assets/c1000cd4-1773-4ead-8eae-6de1912a1b42" />


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the WooCommerce admin pages such as Home, Analytics, Marketing, etc.
2. Confirm that the pages load correctly, especially header, activity panel, and homescreen.
3. Go to `WooCommerce > Home > Add product task
4. Create a new product
5. Confirm `✏How easy was it to add a product?` notice is shown
6. Confirm that the Task list reminder bar is shown on the Product screen. 
7. Go to the embedded pages such as Orders, Products, etc.
8. Confirm that the pages load correctly, especially the header and activity panel.
9. Go to settings > site visibility and confirm that it renders correctly.
10. Smoke-test WC Admin pages to confirm that the changes do not introduce any regressions.

I've also verified the embedded layout renders the same DOM structure as before.

<img width="865" alt="Screenshot 2025-02-25 at 17 16 48" src="https://github.com/user-attachments/assets/c1d38608-4315-4148-91af-1f6a657a799e" />

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Add embedded layout support for WooCommerce admin

- Introduce new embed entry point in webpack config
- Create new embed-specific components for header, layout, and entry point
- Update WCAdminAssets to support embedded page script loading

chihsuan committed:
Refactor WooCommerce Admin client-side architecture

- Migrate client-side code to TypeScript
- Restructure header and layout components
- Extract shared logic into separate files
- Improve type safety and code organization
- Remove deprecated JavaScript files
- Update stylesheets and imports

chihsuan committed:
Refactor header plugin components with improved TypeScript types

- Add type definitions for WooHeaderItem, WooHeaderNavigationItem, and WooHeaderPageTitle
- Simplify component prop types using TypeScript utility types
- Improve type safety for Slot and Fill components
- Remove unnecessary type annotations

chihsuan committed:
Refactor hasTwoColumnLayout function into separate utility file

- Move hasTwoColumnLayout function from layout.js to new utils.tsx file
- Update import paths in related files
- Improve type safety by adding TypeScript type annotations

chihsuan committed:
Add changelog

chihsuan committed:
Add changelog

chihsuan committed:
Fix import path

chihsuan committed:
Add footer

chihsuan committed:
Revert classname

moon0326 commented on plugins/woocommerce/client/admin/client/index.tsx:
```diff
@@ -0,0 +1,57 @@
+
+	// Render the Customer Effort Score Tracks.
+	renderCustomerEffortScoreTracks( appRoot );
+}
```
I think [reactify-classic-payments-settings](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/index.js#L96) check has been removed.

I wasn't able to load WooCommerce -> Settings -> Payments with the feature flag enabled.

moon0326 replied:
Please disregard this comment. It looks like this is implemented in https://github.com/woocommerce/woocommerce/pull/55823

moon0326 commented on plugins/woocommerce/client/admin/client/shared.tsx:
```diff
@@ -0,0 +1,38 @@
+ */
+import { initRemoteLogging } from './lib/init-remote-logging';
+// Initialize remote logging early to log any errors that occur during initialization.
+initRemoteLogging();
```
Since this file is meant to be shared by other components, do you think we should call this method conditionally? Or should we let the upper-level component decide when to call it? Importing a single function from this file will trigger initRemoteLogging, which might be unintentional. However, calling it multiple times wouldn’t be too bad since there is already a [check](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/remote-logging/src/remote-logger.ts#L453) in place.

chihsuan replied:
Good point! I'll move this to upper-level component.

chihsuan replied:
Updated in 1dfaccb337a3d9d9614f4803e409bfa47615e992

moon0326 reviewed (commented):
This is amazing! Nice work 💯

Overall, it looks good and everything works as expected without any errors.

I've left two minor comments.

chihsuan committed:
Add remote logging initialization to client entry points

- Initialize remote logging early in embed, index, and shared entry points
- Ensure consistent error logging across different client entry points

psealock commented on plugins/woocommerce/src/Internal/Admin/WCAdminAssets.php:
```diff
@@ -284,7 +284,7 @@ public function register_scripts() {
 
 		$scripts_map = array(
-			WC_ADMIN_APP    => 'app',
+			WC_ADMIN_APP    => PageController::is_embed_page() ? 'embed' : 'app',
```
Have you done this so that scripts declaring `WC_ADMIN_APP` as a dependency won't trigger loadin of the whole wc-admin, just the embed script?

I'm wondering if there are any scripts out there depending on `WC_ADMIN_APP` and expect the entire app to be present. My guess is no, considering the `wc.*` packages are available on embed pages, but thought maybe worth mentioning.

chihsuan replied:
> Have you done this so that scripts declaring WC_ADMIN_APP as a dependency won't trigger loadin of the whole wc-admin, just the embed script?

Good question! Yes, the refactoring was done with the intention that scripts declaring `WC_ADMIN_APP` as a dependency will only trigger the loading of the necessary embed script rather than the entire `wc-admin`.

> I'm wondering if there are any scripts out there depending on WC_ADMIN_APP and expect the entire app to be present. My guess is no, considering the wc.* packages are available on embed pages, but thought maybe worth mentioning.

I believe there are no scripts that require the entire app to be present on embed pages to function properly. Additionally, this PR extracts the necessary code into the embed script, ensuring compatibility with existing scripts. The other codes are unused, and methods are not exposed globally like `wc.*, so third-party scripts should not be affected.

psealock reviewed (approved):
This is looking really good @chihsuan. The code looks good and I've done a solid amount of poking around. Nothing jumps out, not errors found. Just a question about `WC_ADMIN_APP` script name.

psealock reviewed (approved):
Sounds reasonable to me 🚢

chihsuan committed:
Add back renderPaymentsSettings

chihsuan committed:
Merge branch 'trunk' into enhance/wc-admin-app-embed-page
